packages/jsii/lib/compiler-watch.ts

```diff
@@ -0,0 +1,155 @@
+
+  /**
+   * A promise to the last compilation result achieved when the `.close()` method is called.
+   */
+  private readonly watchComplete: Promise<EmitResult>;
+  /**
+   * This is the "resolve" handler of `this.watchComplete`, and it's actually guaranteed to be set,
+   * and shouldn't be re-written to.
+   */
+  private onWatchEnd!: (lastResult: EmitResult) => void;
+  /**
+   * This is the "reject" handler of `this.watchComplete`, and it's actually guaranteed to be set,
+   * and hsouldn't be re-written to.
+   */
+  private onWatchFailed!: (error: Error) => void;
+  /**
+   * The result of the last completed compilation cycle.
+   */
+  private lastResult?: EmitResult;
+
+  /**
+   * @internal
+   */
+  public constructor(
+    projectInfo: ProjectInfo,
+    configPath: string,
+    compilerOptions: CompilerOptions,
+    onCompilationComplete: TypeScriptProgramHandler,
+    onStatusChanged?: WatchStatusReporter,
+  ) {
+    const host = createWatchCompilerHost(
+      configPath,
+      {
+        ...projectInfo.tsc,
+        ...compilerOptions,
+        noEmitOnError: false,
+      },
+      { ...sys, getCurrentDirectory() { return projectInfo.projectRoot; } },
+      createSemanticDiagnosticsBuilderProgram,
+      undefined, // reportDiagnostics -> We don't pass this because we have our own reporting mechanism
+      onStatusChanged,
+    );
+
+    if (!host.getDefaultLibLocation) {
+      throw new Error('No default library location was found on the TypeScript compiler host!');
+    }
+    const orig = host.afterProgramCreate?.bind(host);
+    host.afterProgramCreate = async builderProgram => {
+      this.lastResult = await onCompilationComplete(builderProgram.getProgram(), host.getDefaultLibLocation!());
+      if (orig != null) { orig(builderProgram); }
+      this.eventEmitter.emit(CompilerWatch.EVENT_COMPILATION_COMPLETE, this.lastResult, this);
+    };
+
+    this.watch = createWatchProgram(host);
+    this.watchComplete = new Promise((resolve, reject) => {
+      this.onWatchEnd = resolve;
+      this.onWatchFailed = reject;
+    });
+  }
+
+  /**
+   * Adds a listener for the 'result' event that is called with EmitResult.
+   *
+   * @param event    the 'result' event
+   * @param listener the listener to be called with the EmitResult.
+   */
+  public on(event: typeof CompilerWatch.EVENT_COMPILATION_COMPLETE, listener: CompilationCompleteHandler): this {
+    this.eventEmitter.on(event, listener);
+    return this;
+  }
+
+  /**
+   * @param timeoutMillis the maximum amount of time to wait for the watch to complete. When
+   *                      provided, the returned promise will reject if it is not resolved by the
+   *                      end of this amount of time.
+   *
+   * @returns a `Promise` that only resolves once the `.close()` method was invoked.
+   */
+  public async block(timeoutMillis?: number): Promise<EmitResult> {
+    const promises = [this.watchComplete];
+
+    if (timeoutMillis != null) {
```

Comment:
why is this needed?

Reply:
This was added safety in case the test never succeeds (because watch is broken?). This is no longer needed.